### ISSUE TITLE
React 18 - How to detect concurrent mode

### ISSUE THREAD

**Issue by incepter** [facebook/react#22964]
Hi,

How to detect that the current used React (stable/release) version in an application supports concurrent mode ? Let's say you are authoring a library that should target several react versions, and only add concurrent features if they are relevant.

I'm doing a [`return typeof React.useSyncExternalStore === "function"`](https://github.com/incepter/react-async-states/blob/main/packages/react-async-states/src/helpers/is-concurrent-mode.js#L4) but I am hesitant about it.

I did create a stackoverflow question about it [here](https://stackoverflow.com/questions/70354232/how-to-detect-react-concurrent-mode)

Thanks in advance

**Comment by anton-g** [comment]
Hi!

If you are building a state management library I _think_ the recommended way is to use the shim for `useSyncExternalStore`: https://www.npmjs.com/package/use-sync-external-store/

You can read about it here: https://github.com/reactwg/react-18/discussions/86

**Comment by incepter** [comment]
Hi @anton-g ; thanks a lot for your reply.

Unfortunately, `useSyncExternalStore` was designed to stop tearing and notify about concurrent writes to a mutable source.

While the library (open source) I am preparing was designed in parallel (I did not integrate the shim yet, and I don't know if the library needs it), my problem is more about providing an API to `trigger` the component suspension.
The challenge is to not have two versions of the library (one for react18+ and the other for older versions); so far, I came with a `read()` function (see implementation bellow) like all the examples of suspense that suspends if it detects that you are using concurrent mode throws.

The thing is; the `isConcurrentMode()` i am using scans statically the react module and checks for a property and not at runtime when the app can be mounted using `render` or `createRoot`; 

This puts an extra burden on the developers shoulders to keep these stuff in sync, if the render method was used; the thrown suspender will bubbles up to the nearest error boundary; which is not the excepted behavior. This could get more complicated if the library supports an extra boolean to tune this.
But in general I think opting towards the read function means that you want to suspend a component, at least for now.

Implementation of read function:

```javascript
returnValue.read = function readInConcurrentMode() {
    if (enableComponentSuspension) {
      if (isConcurrentMode()) {
        if (AsyncStateStatus.pending === asyncState?.currentState?.status && asyncState.suspender) {
          throw asyncState.suspender;
        }
      } else {
        if (__DEV__ && !didWarnAboutUnsupportedConcurrentFeatures) {
          console.error("[Warning] You are calling useAsyncState().read() without having react 18 or above " +
            "if the library throws, you will get an error in your app. You will be receiving the state value without " +
            "any suspension. Please consider upgrading to react 18 or above to use concurrent features.")
          didWarnAboutUnsupportedConcurrentFeatures = true;
        }
      }
    }
    return stateValue;
  }
```

**Comment by eps1lon** [comment]
Could you update the issue title to reflect that is specifically about an API to explicitly suspend a component?

**Comment by incepter** [comment]
Done.

This issue is about detecting whether the used react's version supports concurrent features that the library offers, like suspending components.

**Comment by eps1lon** [comment]
Since the API about how to suspend (e.g. throw promise) it's really about a specific API to suspend. 

The reason why "API how to detect concurrent features" is not actionable is that the next question will always be "what do you want to do?". So it's better to make the issue about the what and not the how.

**Comment by gaearon** [comment]
I don't quite understand what is being asked.

The "throw Promise" API is not "officially" supported and will likely be deprecated in favor of a stable API for that once one exists. A stable API for this feature does not currently exist.

However, "throw Promise" has worked since some React 16.x version. It's not new in React 18. So I'm not sure why detecting React 18 is relevant.

**Comment by incepter** [comment]
> I don't quite understand what is being asked.
>

My original question was if there is a way to detect the react current version capabilities like component suspending to have a consistent behavior in some designed api.

> The "throw Promise" API is not "officially" supported and will likely be deprecated in favor of a stable API for that once one exists. A stable API for this feature does not currently exist.
>

Ok. This settles it. Hopefully it will be something more useful then.

But, aren't we supposed to throw promises to suspend in react 18 ?

> However, "throw Promise" has worked since some React 16.x version. It's not new in React 18. So I'm not sure why detecting React 18 is relevant.
>

I didn't know of any support for it in react (except through experimental releases).

It is not relevant, it is just to warn the developer about using an api designed to work with another version of react than his.

**Comment by gaearon** [comment]
>But, aren't we supposed to throw promises to suspend in react 18 ?

Yes, but the API is in an awkward grey area. In 18.0, the support for Suspense for data fetching is [mostly for opinionated frameworks](https://reactjs.org/blog/2022/03/29/react-v18.html#suspense-in-data-frameworks). We think there's a few more things we need to provide before we'd recommend to use it more broadly than in frameworks. I understand it’s not a satisfying answer, but we just don’t have all the pieces in place yet, and a stable API would likely be a bit different.

>It is not relevant, it is just to warn the developer about using an api designed to work with another version of react than his.

I don’t remember what’s the first release to support it, so you’d need to find that via trial and error. I suspect that 16.7+ supports it. If your library uses Hooks it’s already 16.8+ so you shouldn’t need to check anything.

**Comment by incepter** [comment]
> I understand it’s not a satisfying answer, but we just don’t have all the pieces in place yet, and a stable API would likely be a bit different.
>

On the opposite. I understand better now. Let's keep track of how things will evolve then.

Thanks for the explanations Dan.
